File: selected_coding_interview/codes/python/lc_3_longest_substring_without_repeating_characters_s2.py
"""
File: lc_3_longest_substring_without_repeating_characters_s2.py
Created Time: 2025-12-30
Author: krahets
"""

from include import *


# ===== Solution Code =====
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        dic = {}
        res = tmp = 0
        for j in range(len(s)):
            i = dic.get(s[j], -1) # 获取索引 i
            dic[s[j]] = j # 更新哈希表
            tmp = tmp + 1 if tmp < j - i else j - i # dp[j - 1] -> dp[j]
            res = max(res, tmp) # max(dp[j - 1], dp[j])
        return res


# ======= Test Case =======
# Test case 1
test_input = "example"

# ====== Driver Code ======
slt = Solution()
result = slt.lengthOfLongestSubstring(test_input)
print(result)

File: selected_coding_interview/codes/java/lc_3_longest_substring_without_repeating_characters/lc_3_longest_substring_without_repeating_characters_s2.java
/*
* File: lc_3_longest_substring_without_repeating_characters_s2.java
* Created Time: 2025-12-30
* Author: krahets
*/

package lc_3_longest_substring_without_repeating_characters;

import include.*;
import java.util.*;

// ===== Solution Code =====
    class Solution {
        public int lengthOfLongestSubstring(String s) {
            Map<Character, Integer> dic = new HashMap<>();
            int res = 0, tmp = 0, len = s.length();
            for(int j = 0; j < len; j++) {
                int i = dic.getOrDefault(s.charAt(j), -1); // 获取索引 i
                dic.put(s.charAt(j), j); // 更新哈希表
                tmp = tmp < j - i ? tmp + 1 : j - i; // dp[j - 1] -> dp[j]
                res = Math.max(res, tmp); // max(dp[j - 1], dp[j])
            }
            return res;
        }
    }

public class lc_3_longest_substring_without_repeating_characters_s2 {
    public static void main(String[] args) {
        // ======= Test Case =======
        // Test case 1
        String test_input = "example";

        // ====== Driver Code ======
        Solution slt = new Solution();
        var result = slt.lengthOfLongestSubstring(test_input);
        System.out.println(result);

    }
}
